core/drivers/crypto/caam/caam_jr.c

```diff
@@ -0,0 +1,676 @@
+ * (@wait_job_ids parameter)
+ *
+ * @wait_job_ids  Expected Jobs to be complete
+ */
+static uint32_t do_jr_dequeue(uint32_t wait_job_ids)
+{
+	uint32_t ret_job_id = 0;
+	struct caller_info *caller = NULL;
+	struct outring_entry *jr_out = NULL;
+	struct caam_jobctx *jobctx = NULL;
+	uint32_t exceptions = 0;
+	bool found = false;
+	uint16_t idx_jr = 0;
+	uint16_t nb_jobs_done = 0;
+	size_t nb_jobs_inv = 0;
+
+	exceptions = cpu_spin_lock_xsave(&jr_privdata->outlock);
+
+	nb_jobs_done = caam_hal_jr_get_nbjob_done(jr_privdata->baseaddr);
+
+	if (nb_jobs_done == 0) {
+		cpu_spin_unlock_xrestore(&jr_privdata->outlock, exceptions);
+		return ret_job_id;
+	}
+
+	/* Ensure that output ring descriptor entries are not in cache */
+	if ((jr_privdata->outread_index + nb_jobs_done) >
+	    jr_privdata->nb_jobs) {
+		/*
+		 * Invalidate the whole circular job buffer because some
+		 * completed job rings are at the beginning of the buffer
+		 */
+		jr_out = jr_privdata->outrings;
+		nb_jobs_inv = jr_privdata->nb_jobs;
+	} else {
+		/* Invalidate only the completed job */
+		jr_out = &jr_privdata->outrings[jr_privdata->outread_index];
+		nb_jobs_inv = nb_jobs_done;
+	}
+
+	cache_operation(TEE_CACHEINVALIDATE, jr_out,
+			sizeof(struct outring_entry) * nb_jobs_inv);
+
+	do {
+		jr_out = &jr_privdata->outrings[jr_privdata->outread_index];
+
+		/*
+		 * Lock the caller information array because enqueue is
+		 * also touching it
+		 */
+		cpu_spin_lock(&jr_privdata->callers_lock);
+		for (idx_jr = 0, found = false; idx_jr < jr_privdata->nb_jobs;
+		     idx_jr++) {
+			/*
+			 * Search for the caller information corresponding to
+			 * the completed JR.
+			 * Don't use the outread_index or inwrite_index because
+			 * completion can be out of order compared to input
+			 * buffer
+			 */
+			caller = &jr_privdata->callers[idx_jr];
+			if (caam_desc_pop(&jr_out->desc) == caller->pdesc) {
+				jobctx = caller->jobctx;
+				jobctx->status =
+					caam_read_jobstatus(&jr_out->status);
+
+				/* Update return Job IDs mask */
+				if (caller->job_id & wait_job_ids)
+					ret_job_id |= caller->job_id;
+
+				JR_TRACE("JR id=%d, context @0x%08" PRIxVA,
+					 caller->job_id, (vaddr_t)jobctx);
+				/* Clear the Entry Descriptor DMA */
+				caller->pdesc = 0;
+				caller->job_id = JR_JOB_FREE;
+				found = true;
+				JR_TRACE("Free space #%d in the callers array",
+					 idx_jr);
+				break;
+			}
+		}
+		cpu_spin_unlock(&jr_privdata->callers_lock);
+
+		/*
+		 * Remove the JR from the output list even if no
+		 * JR caller found
+		 */
+		caam_hal_jr_del_job(jr_privdata->baseaddr);
+
+		/*
+		 * Increment index to next JR output entry taking care that
+		 * it is a circular buffer of nb_jobs size.
+		 */
+		jr_privdata->outread_index++;
+		jr_privdata->outread_index %= jr_privdata->nb_jobs;
+
+		if (found && jobctx->callback) {
+			/* Finally, execute user's callback */
+			jobctx->callback(jobctx);
+		}
+
+	} while (--nb_jobs_done);
```

Comment:
prefer `nb_jobs_done--`

Reply:
ok

Reply:
In fact change to a for loop because this change it's breaking to execution.